src/main/java/org/jenkinsci/plugins/workflow/steps/durable_task/DurableTaskStep.java

```diff
@@ -417,6 +426,50 @@ private void getWorkspaceProblem(Exception x) {
             return l;
         }
 
+        /**
+         * Interprets {@link PrintStream#close} as a signal to end a final newline if necessary.
+         */
+        private static final class NewlineSafeTaskListener implements TaskListener {
+
+            private static final long serialVersionUID = 1;
+
+            private final TaskListener delegate;
+            private transient PrintStream logger;
+
+            NewlineSafeTaskListener(TaskListener delegate) {
+                this.delegate = delegate;
+            }
+
+            // Similar to DecoratedTaskListener:
+            @Override public synchronized PrintStream getLogger() {
+                if (logger == null) {
+                    LOGGER.fine("creating filtered stream");
+                    OutputStream base = delegate.getLogger();
+                    OutputStream filtered = new FilterOutputStream(base) {
+                        boolean nl = true; // empty string does not need a newline
+                        @Override public void write(int b) throws IOException {
```

Comment:
Is this in a context where performance matters? If so, maybe good to override the other `write` overloads, especially since I think just looking at the last byte would still work fine for bulk writes.

Reply:
True, overloading `write(byte[], int, int)` could be slightly faster.